staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/atomic.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func validateAtomic(schema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+
+	if schema == nil {
+		errs = append(errs, field.Required(fldPath, "cannot be empty as part of an atomic type"))
+		return errs
+	}
+
+	switch schema.Type {
+	case "":
+		if !schema.XIntOrString {
+			errs = append(errs, field.Required(fldPath.Child("type"), "cannot be empty as part of an atomic type"))
+		}
+	case "object":
+		if schema.XMapType == nil || *schema.XMapType != "atomic" { // default is granular
+			errs = append(errs, field.Invalid(fldPath.Child("x-kubernetes-map-type"), schema.XMapType, "must be atomic as part of an atomic type"))
+		} else if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
```

Comment:
Shouldn't the whole object be atomic? rather than all of it's members. And I don't think we have to keep recursing to find atomicity everywhere?

Reply:
From SMD perspective, we consider something atomic as soon as we see one.

Reply:
that was my question when I asked whether it is recursive. But I see what you mean: if you patch, you start at the root and merge maps/lists until you find one which is atomic. Then you replace and stop. In that sense it applies to the full scope.

Reply:
But I am not sure I would call everything in the scope atomic then. It is merely a consequence of recursive-until-atomic-then-stop in SSA, but one could imagine a patch operator which first recurses deep into an object, skipping atomic levels and then do something non-atomically.